File: couchbase-entity/src/main/java/com/kaufland/model/field/CblFieldHolder.java
package com.kaufland.model.field;

import com.kaufland.ElementMetaModel;
import com.kaufland.generation.TypeConversionMethodsGeneration;
import com.kaufland.util.TypeUtil;
import com.squareup.javapoet.ClassName;
import com.squareup.javapoet.CodeBlock;
import com.squareup.javapoet.FieldSpec;
import com.squareup.javapoet.MethodSpec;
import com.squareup.javapoet.TypeName;
import com.thoughtworks.qdox.model.JavaClass;
import com.thoughtworks.qdox.model.JavaField;
import com.thoughtworks.qdox.model.JavaType;
import com.thoughtworks.qdox.model.impl.DefaultJavaParameterizedType;

import org.apache.commons.lang3.StringUtils;
import org.apache.commons.lang3.text.WordUtils;

import java.util.Collections;
import java.util.List;

import javax.lang.model.element.Element;
import javax.lang.model.element.Modifier;

import kaufland.com.coachbasebinderapi.DefaultValue;
import kaufland.com.coachbasebinderapi.Field;

public class CblFieldHolder extends CblBaseFieldHolder {

    private String subEntityPackage;

    private String subEntitySimpleName;

    private boolean subEntityIsTypeParam;

    private boolean isIterable;

    public CblFieldHolder(Field field, ElementMetaModel metaModel) {
        super(field.name(), field);

        String typeName = getTypeMirror().toString();
        if (metaModel.isMapWrapper(typeName)) {

            subEntitySimpleName = TypeUtil.getSimpleName(getTypeMirror()) + "Wrapper";
            subEntityPackage = TypeUtil.getPackage(getTypeMirror());
            subEntityIsTypeParam = field.list();

        }
        if (field.list()) {
            isIterable = true;
        }
    }

    public String getSubEntitySimpleName() {
        return subEntitySimpleName;
    }

    public TypeName getSubEntityTypeName() {
        return ClassName.get(subEntityPackage, subEntitySimpleName);
    }

    public boolean isSubEntityIsTypeParam() {
        return subEntityIsTypeParam;
    }



    public boolean isTypeOfSubEntity() {
        return !StringUtils.isBlank(subEntitySimpleName);
    }

    @Override
    public MethodSpec getter(String dbName, boolean useMDocChanges) {
        TypeName returnType = TypeUtil.parseMetaType(getTypeMirror(), isIterable, getSubEntitySimpleName());

        MethodSpec.Builder builder = MethodSpec.methodBuilder("get" + accessorSuffix()).
                addModifiers(Modifier.PUBLIC).
                returns(returnType);


        if (isTypeOfSubEntity()) {
            returnType = TypeUtil.parseMetaType(getTypeMirror(), isIterable, getSubEntitySimpleName());
            TypeName castType = isSubEntityIsTypeParam() ? TypeUtil.createListWithMapStringObject() : TypeUtil.createMapStringObject();

            if (useMDocChanges) {
                builder.addCode(CodeBlock.builder().beginControlFlow("if(mDocChanges.containsKey($N))", getConstantName()).
                        addStatement("return ($T) $T.fromMap(($T)mDocChanges.get($N))", returnType, getSubEntityTypeName(), castType, getConstantName()).
                        endControlFlow().
                        build());
            }
            builder.addCode(CodeBlock.builder().beginControlFlow("if(mDoc.containsKey($N))", getConstantName()).
                    addStatement("return ($T) $T.fromMap(($T)mDoc.get($N))", returnType, getSubEntityTypeName(), castType, getConstantName()).
                    endControlFlow().
                    build());

            builder.addStatement("return null");
        } else {

            TypeName forTypeConversion = evaluateClazzForTypeConversion();
            if (useMDocChanges) {
                builder.addCode(CodeBlock.builder().beginControlFlow("if(mDocChanges.containsKey($N))", getConstantName()).
                        addStatement("return " + TypeConversionMethodsGeneration.READ_METHOD_NAME + "(mDocChanges.get($N), $T.class)", getConstantName(), forTypeConversion).
                        endControlFlow().
                        build());
            }

            builder.addCode(CodeBlock.builder().beginControlFlow("if(mDoc.containsKey($N))", getConstantName()).
                    addStatement("return " + TypeConversionMethodsGeneration.READ_METHOD_NAME + "(mDoc.get($N), $T.class)", getConstantName(), forTypeConversion).
                    endControlFlow().
                    build());

            if (isDefault()) {
                builder.addStatement("return " + TypeConversionMethodsGeneration.READ_METHOD_NAME + "(mDocDefaults.get($N), $T.class)", getConstantName(), forTypeConversion);
            } else if(forTypeConversion.isPrimitive()){
                builder.addStatement("return $L.get($T.class)", DefaultValue.class.getCanonicalName(), forTypeConversion);
            } else {
                builder.addStatement("return null");
            }

        }

        return builder.build();
    }


    @Override
    public MethodSpec setter(String dbName, TypeName entityTypeName, boolean useMDocChanges) {
        TypeName fieldType = TypeUtil.parseMetaType(getTypeMirror(),isIterable, getSubEntitySimpleName());
        MethodSpec.Builder builder = MethodSpec.methodBuilder("set" + accessorSuffix()).
                addModifiers(Modifier.PUBLIC).
                addParameter(fieldType, "value").
                returns(entityTypeName);

        String docName = useMDocChanges ? "mDocChanges" : "mDoc";

        if (isTypeOfSubEntity()) {
            builder.addStatement("$N.put($N, $T.toMap(($T)value))", docName, getConstantName(), getSubEntityTypeName(), fieldType);
            builder.addStatement("return this");
        } else {
            TypeName forTypeConversion = evaluateClazzForTypeConversion();
            builder.addStatement("$N.put($N, " + TypeConversionMethodsGeneration.WRITE_METHOD_NAME + "(value, $T.class))", docName, getConstantName(), forTypeConversion);
            builder.addStatement("return this");
        }

        return builder.build();
    }

    @Override
    public List<FieldSpec> createFieldConstant() {

        FieldSpec fieldAccessorConstant = FieldSpec.builder(String.class, getConstantName(), Modifier.FINAL, Modifier.PUBLIC, Modifier.STATIC).
                initializer("$S", getDbField()).
                build();

        return Collections.singletonList(fieldAccessorConstant);
    }

    private TypeName evaluateClazzForTypeConversion() {
        if (isIterable) {
            return TypeName.get(List.class);
        }

        return TypeUtil.parseMetaType(getTypeMirror(),isIterable, getSubEntitySimpleName());
    }
}

File: couchbase-entity/src/main/java/com/kaufland/model/field/CblFieldHolder.kt
package com.kaufland.model.field

import com.kaufland.generation.model.KDocGeneration
import com.kaufland.generation.model.TypeConversionMethodsGeneration
import com.kaufland.model.deprecated.DeprecatedModel
import com.kaufland.util.TypeUtil
import com.squareup.kotlinpoet.*
import kaufland.com.coachbasebinderapi.Field
import org.apache.commons.lang3.StringUtils

class CblFieldHolder(field: Field, allWrappers: List<String>) : CblBaseFieldHolder(field.name, field) {

    private var subEntityPackage: String? = null

    var subEntitySimpleName: String? = null
        private set

    var isSubEntityIsTypeParam: Boolean = false
        private set

    override var isIterable: Boolean = false

    val subEntityTypeName: TypeName
        get() = ClassName(subEntityPackage!!, subEntitySimpleName!!)

    val isTypeOfSubEntity: Boolean
        get() = !StringUtils.isBlank(subEntitySimpleName)

    override val fieldType: TypeName = TypeUtil.parseMetaType(typeMirror, isIterable, subEntitySimpleName)

    init {
        if (allWrappers.contains(typeMirror.toString())) {

            subEntitySimpleName = TypeUtil.getSimpleName(typeMirror) + "Wrapper"
            subEntityPackage = TypeUtil.getPackage(typeMirror)
            isSubEntityIsTypeParam = field.list
        }
        if (field.list) {
            isIterable = true
        }
    }

    override fun interfaceProperty(isOverride: Boolean): PropertySpec {
        val returnType = TypeUtil.parseMetaType(typeMirror, isIterable, subEntitySimpleName).copy(nullable = true)
        val modifiers = listOfNotNull(KModifier.PUBLIC, KModifier.OVERRIDE.takeIf { isOverride })
        return PropertySpec.builder(accessorSuffix(), returnType.copy(true), modifiers).mutable(true).build()
    }

    override fun property(dbName: String?, possibleOverrides: Set<String>, useMDocChanges: Boolean, deprecated: DeprecatedModel?): PropertySpec {
        val returnType = TypeUtil.parseMetaType(typeMirror, isIterable, subEntitySimpleName).copy(nullable = true)

        val propertyBuilder = PropertySpec.builder(accessorSuffix(), returnType.copy(true), KModifier.PUBLIC, KModifier.OVERRIDE).mutable(true)

        val getter = FunSpec.getterBuilder()
        val setter = FunSpec.setterBuilder().addParameter("value", String::class)

        deprecated?.addDeprecated(dbField, propertyBuilder)

        val docName = if (useMDocChanges) "mDocChanges" else "mDoc"

        if (isTypeOfSubEntity) {
            val castType = if (isSubEntityIsTypeParam) TypeUtil.listWithMutableMapStringAnyNullable() else TypeUtil.mutableMapStringAnyNullable()

            if (useMDocChanges) {
                getter.addCode(CodeBlock.builder().beginControlFlow("if(mDocChanges.containsKey(%N))", constantName).addStatement("return·%T.fromMap(mDocChanges.get(%N) as? %T)", subEntityTypeName, constantName, castType).endControlFlow().build())
            }
            getter.addCode(CodeBlock.builder().beginControlFlow("if(mDoc.containsKey(%N))", constantName)
                    /** In case the key for the subentity is set but the value is null it would result in a classcastexception since null can't be cast to any type*/
                    .addStatement("return·%T.fromMap(mDoc.get(%N) as? %T)", subEntityTypeName, constantName, castType).endControlFlow().build())

            getter.addStatement("return null")

            setter.addStatement("%N.put(%N, %T.toMap(value))", docName, constantName, subEntityTypeName)
        } else {

            val forTypeConversion = evaluateClazzForTypeConversion()
            if (useMDocChanges) {
                getter.addCode(CodeBlock.builder().beginControlFlow("if(mDocChanges.containsKey(%N))", constantName).addStatement("return " + TypeConversionMethodsGeneration.READ_METHOD_NAME + "(mDocChanges.get(%N), %T::class)", constantName, forTypeConversion).endControlFlow().build())
            }

            getter.addCode(CodeBlock.builder().beginControlFlow("if(mDoc.containsKey(%N))", constantName).addStatement("return " + TypeConversionMethodsGeneration.READ_METHOD_NAME + "(mDoc.get(%N), %T::class)", constantName, forTypeConversion).endControlFlow().build())

            getter.addStatement("return null")

            setter.addStatement("%N.put(%N, " + TypeConversionMethodsGeneration.WRITE_METHOD_NAME + "(value, %T::class))", docName, constantName, forTypeConversion)
        }

        if (comment.isNotEmpty()) {
            propertyBuilder.addKdoc(KDocGeneration.generate(comment))
        }

        return propertyBuilder.setter(setter.build()).getter(getter.build()).build()
    }

    fun ensureType(resultType: TypeName, format: String, vararg args: Any?): CodeBlock {
        val forTypeConversion = evaluateClazzForTypeConversion()
        return CodeBlock.of("${TypeConversionMethodsGeneration.WRITE_METHOD_NAME}<%T>($format, %T::class)", resultType, *args, forTypeConversion)
    }

    override fun builderSetter(dbName: String?, packageName: String, entitySimpleName: String, useMDocChanges: Boolean, deprecated: DeprecatedModel?): FunSpec {
        val fieldType = TypeUtil.parseMetaType(typeMirror, isIterable, subEntitySimpleName)
        val builder = FunSpec.builder("set" + accessorSuffix().capitalize()).addModifiers(KModifier.PUBLIC).addParameter("value", fieldType).returns(ClassName(packageName, "$entitySimpleName.Builder"))

        if (this.comment.isNotEmpty()) {
            builder.addKdoc(KDocGeneration.generate(comment))
        }

        if (deprecated?.evaluateFieldDeprecationLevel(dbField) == DeprecationLevel.ERROR && deprecated.addDeprecated(dbField, builder)) {
            builder.addStatement("throw %T()", UnsupportedOperationException::class)
        } else {
            builder.addStatement("obj.${accessorSuffix()} = value")
            builder.addStatement("return this")
        }

        return builder.build()
    }

    override fun createFieldConstant(): List<PropertySpec> {
        val fieldAccessorConstant = PropertySpec.builder(constantName, String::class, KModifier.FINAL, KModifier.PUBLIC).initializer("%S", dbField).addAnnotation(JvmField::class).build()
        return listOf(fieldAccessorConstant)
    }

    fun evaluateClazzForTypeConversion(): TypeName {
        return if (isIterable) {
            if (TypeUtil.isMap(fieldType)) {
                TypeUtil.string()
            } else {
                fieldType
            }
        } else TypeUtil.parseMetaType(typeMirror, isIterable, false, subEntitySimpleName)
    }
}
